### DISCUSSION TITLE
Orjson in Pydantic V2

### DISCUSSION THREAD

**Discussion by phillipuniverse** [pydantic/pydantic#6388] [Pydantic V2]
In Pydantic V1, I use `orjson` for my `json_dumps` and `json_loads`:

```python
from pydantic import BaseModel
import orjson

def _orjson_dumps(v: Any, *, default: Any) -> str:
    # orjson.dumps returns bytes, to match standard json.dumps we need to decode
    # comes from https://docs.pydantic.dev/usage/exporting_models/#custom-json-deserialisation
    return orjson.dumps(
        v,
        default=default,
    ).decode()


class MyModel(BaseModel):

    class Config:
        json_dumps = _orjson_dumps
        json_loads = orjson.loads

```

I am having a little trouble figuring out how to do this in Pydantic V2, or if this is even necessary. With pydantic-core rewritten in Rust and the performance improvements along with it, is using orjson even necessary? Or would the configuration look more like implementing a `SchemaSerializer`? Thinking something like this in V2, but wondering what I would give up in Pydantic V2 for this somewhat-ham-fisted approach:

```python
from pydantic_core import SchemaSerializer

class OrJsonSerializer(SchemaSerializer):

    def to_json(self, *args, **kwargs):
        return orjson.dumps(...) # implementation not complete, need to understand args/kwargs

    def to_python(self, *args, **kwargs): # real signature unknown
        return orjson.loads(...). # implementation not complete, need to understand args/kwargs

class MyModel(BaseModel):

    __pydantic_serializer__ = OrJsonSerializer()
```

**Comment by adriangb** [comment]
> With pydantic-core rewritten in Rust and the performance improvements along with it, is using orjson even necessary?

No, using `orjson` is generally not necessary. I'm not sure that we've benchmarked it but the built-in json dumping/loading should be fast enough.

**Reply by shurshilov** [reply]
Orjson give very good perfomance for example in FASTAPI 2 codes
This with orjson
![image](https://github.com/pydantic/pydantic/assets/11828278/8e7da8c9-32d5-45cc-8136-886ab9b11d73)
![image](https://github.com/pydantic/pydantic/assets/11828278/c66683f4-6c6a-489f-a337-23392d0ac5bd)

And this without
![image](https://github.com/pydantic/pydantic/assets/11828278/888d2c63-2ef6-4ac2-be53-f16ebcdc4df9)
![image](https://github.com/pydantic/pydantic/assets/11828278/2b8241a9-56f2-4541-9dab-97a98c737376)
810 vs 583 resuests/second
892 vs 643 MB  per 10seconds
~20% speedup
So... how use orjson inside pydantic?
Because if we use this 
![image](https://github.com/pydantic/pydantic/assets/11828278/600b5e48-a60b-40da-adfd-0d721d57099d)
![image](https://github.com/pydantic/pydantic/assets/11828278/c64ba844-6acc-43e2-9ae6-d7027a0eec9f)

I think pydantic use slowly serelize by default...
P.S.
tools for testing _autocannon 127.0.0.1:81 -d 10 -c 30 -w 3_
start fastapi uvicorn 1 worker

**Reply by samuelcolvin** [reply]
Full answer below which shows that pydantic's json parsing is generally faster than orjson would be https://github.com/pydantic/pydantic/discussions/6388#discussioncomment-13944196.

**Comment by radugrosu** [comment]
According to the [Migration Guide](https://docs.pydantic.dev/latest/migration/#changes-to-config):

> The following config settings have been removed:
> * json_loads
> * json_dumps

however, no alternative is suggested, and no explanation of why, or to what effect. It seems like a very strange decision, and one that probably puts people off from upgrading from V1.

**Reply by samuelcolvin** [reply]
Pydantic V2 parses and serialises JSON directly in Rust as an integrated part of validation and marshelling, this is explained at length in many places including the V2 roadmap article.

Because of this we can't and don't need to offer integration with other JSON libraries.

The rate of adoption of pydantic V2 suggests this hasn't put people off, probably because many of them took the time to do some research and think for a second.

**Reply by radugrosu** [reply]
I'm not usually being told off by maintainers on GitHub, but when I do, I go [msgspec](https://gist.github.com/jcrist/d62f450594164d284fbea957fd48b743).

**Reply by arthur-st** [reply]
@samuelcolvin It's good to hear that this information is clearly explained in many places, but, unfortunately, it that is not the case for any of the following places:

- [Documentation page for JSON](https://docs.pydantic.dev/latest/concepts/json/)
- [Documentation page for serialization](https://docs.pydantic.dev/latest/concepts/serialization/#modelmodel_dump_json)

**Reply by harshit98** [reply]
@radugrosu 

>_however, no alternative is suggested, and no explanation of why, or to what effect. It seems like a very strange decision, and one that probably puts people off from upgrading from V1._

Agreed. I was also planning to migrate from Pydantic V1 to V2. We use msgspec with Pydantic V1 for JSON handling. After going through the migration guide, I realised that we can't use any custom JSON handler with Pydantic V2 now. I can't trade off over JSON performance.

**Comment by shurshilov** [comment]
> Pydantic V2 parses and serialises JSON directly in Rust as an integrated part of validation and marshelling, this is explained at length in many places including the V2 roadmap article.
> 
> Because of this we can't and don't need to offer integration with other JSON libraries.
> 
> The rate of adoption of pydantic V2 suggests this hasn't put people off, probably because many of them took the time to do some research and think for a second.

it doesn't matter to everyone what it is written on, even rust, even c++, as a library user in python, I want speed. Your rust is the same speed as Python json. What is the meaning of it for me as a user? besides the fact that it is written in RUST))kkak it's cool RUST - no)

**Reply by vadim-su** [reply]
Python is not about speed of execution. It's more about programming (prototyping) speed. If you need execution speed, you may need a different programming language.

**Reply by UnoYakshi** [reply]
@vadim-su, the fact that Python is one of the most expressive languages and is great for prototyping doesn't mean it can't be fast[er]. Where would we be w/o C interop (hello, numpy)?

**Comment by shurshilov** [comment]
Any highly loaded professional project removes the pydantic as it is slow. What is not clear?

**Comment by HansBrende** [comment]
@samuelcolvin Python's built-in `json.loads` in combination with `model_validate` appears to be roughly **45% faster** than `model_validate_json`. Running on pydantic 2.5.3. 

Approach 1: 15 seconds (combined total over 210 server responses)
```python
result = MyModel[MyProps](**json.loads(body_bytes))
```

Approach 2: 22 seconds (combined total over the same 210 server responses)
```python
result = MyModel[MyProps].model_validate_json(body_bytes)
```

Both results are better than V1, however, I'm struggling to see the promised speed advantage of `model_validate_json`... as it seems to be in fact... slower... than `json.loads`.

**Reply by HansBrende** [reply]
(Addendum)

Approach 3: 13 seconds
```python
result = MyModel[MyProps](**orjson.loads(body_bytes))
```

**Reply by adriangb** [reply]
Can you give a reproducible code snippet? I don’t know what the contents of MyModel or body_bytes are

**Reply by HansBrende** [reply]
@adriangb Unfortunately I can't give you that information as it's proprietary. I can tell you that `body_bytes` is a `bytes` object and `MyModel` has the shape `results: list[MyResultModel[MyProps]]`, where `MyResultModel` has the shape `id: int, uuid: UUID, properties: MyProps`.

> I'm not sure that we've benchmarked it but the built-in json dumping/loading should be fast enough.

Maybe a good first step would be to actually write a benchmark?

**Reply by HansBrende** [reply]
(Addendum `#2`)

Approach 4: **4 seconds** (after swapping out `BaseModel` for `msgspec.Struct` in my model definitions)

```python
result = msgspec.json.decode(body_bytes, type=MyModel[MyProps])
```

I hope pydantic can improve the speed here at some point (and I may migrate back at that point), but for now it's clear to me what I need to do to get the necessary performance metrics out of this critical section of my application. @adriangb hope to see some benchmarks from Pydantic at some point... I'll keep an eye out!

**Reply by phillipuniverse** [reply]
@HansBrende surely you can provide some general information if you can’t provide specifics? Like maybe how many bytes body_bytes is, or how many key/value pairs, or even an obfuscated representation of the JSON

**Reply by adriangb** [reply]
Agreed, I’m not sure what you’d like benchmarked. I could benchmark something that’s 1000% faster than `json.loads` but we’re still slower for your use case. It’s not like there are no benchmarks: https://github.com/pydantic/jiter. Benchmarks are difficult to do fairly, I’d rather profile your use case, see what’s slow and improve that vs compare to other libraries.

what matters is the model shape (recursion depth, objects vs arrays, etc) you should be able to provide a version that can be shared publicly.

**Reply by HansBrende** [reply]
@phillipuniverse sure. A total of 626,471,506 bytes are decoded over 210 concurrent server responses, giving an average of 2,983,197.6 bytes per response. (Obviously, I only start measuring the time passed after I've received a response and have a bytes object... these 210 timings are then added together for the total time spent decoding).

And here is the full shape of the data:

```
Response[P]:
    results: list[Result[P]]

Result[P]:
    id: int
    uuid: UUID
    properties: P
```
The properties type `P` in this case looks something like the following:

```
P:
    field1: str
    field2: str
    field3: Optional[int] = None
    field4: Optional[int] = None
    field5: Optional[int] = None
    field6: Optional[int] = None
    field7: Optional[int] = None
    field8: Optional[uuid.UUID] = None
    field9: Optional[uuid.UUID] = None
    field10: Optional[bool] = None
    field11: Optional[int] = None
    field12: Optional[str] = None
    field13: Optional[str] = None
    field14: Optional[bool] = None
    field15: IdModel
        id: string
        uuid: UUID
    field16: Optional[str] = None
    field17: Optional[float] = None
    field18: Optional[date] = None
```

Note: there are fields in the JSON that I am ignoring, thus don't have properties listed here to avoid processing them.

**Reply by HansBrende** [reply]
@adriangb as you can see from above, the recursion depth is _fairly_ low. There are a _lot_ of results however (I think the max number of results I saw from one of the responses was around 3000). I'd estimate that more than half the total bytes are made up of JSON fields that I am ignoring. (But the ignored fields do not increase the recursion depth any further than the `field15: IdModel` does).

**Reply by samuelcolvin** [reply]
Thanks so much @HansBrende for providing this, with your pseudo code I was able to reproduce your finding of `model_validate(json.loads(...))` being faster than `model_validate_json(...)`, gist of my script [here](https://gist.github.com/samuelcolvin/0b7e08f6358d602bb3f7b74b7109fc6e).

I created a flamegraph of the `model_validate_json` call (firefox profiler view [here](https://profiler.firefox.com/public/c1vqeskfj0985w5zv9x6hcrn11ccxhm8v5b4290/flame-graph/?globalTrackOrder=0&symbolServer=http%3A%2F%2F127.0.0.1%3A3000%2F30q3xxy0a9jfkc2g1npamlm91mn2aq6agzcm0da&thread=0&v=10)), and was able to see that around 16% of the time was being spend in [`LazyIndexMap.get`](https://github.com/pydantic/jiter/blob/1ce40b322ae29fb4788ba03b476a0e72cb9ce3ef/src/lazy_index_map.rs#L56). Basically before looking up any keys in a JSON object, we were constructing a `HashMap`.

Constructing this `HashMap` isn't really necessary for objects with only a few keys - it's faster to iterate over the original vec directly looking for a match, I've started implementing that in https://github.com/pydantic/jiter/pull/55.

With that performance of `model_validate_json` exceeds `model_validate`:

### pydantic main + pydantic-core 2.14.6

```
model_validate_json(...): 5.31ms
model_validate(json.loads(...)): 5.21ms
json.loads(...): 1.80ms
pydantic_core.from_json(...): 1.32ms
```

### pydantic main + pydantic-core main + jiter https://github.com/pydantic/jiter/pull/55

```
model_validate_json(...): 4.61ms
model_validate(json.loads(...)): 5.25ms
json.loads(...): 1.76ms
pydantic_core.from_json(...): 1.32ms
```

_(note, this performance will improve when released as the compile here didn't use PGO)_

---

The good news more generally is that pydantic's performance should improve significantly (around 30% in most cases) when we merge https://github.com/pydantic/pydantic-core/pull/1085 - this uses the new PyO3 API, but we can't merge that code until PyO3 releases a beta to avoid breaking some distributions (mostly linux distros) that build pydantic-core from source using crates.io.

Although the PyO3 changes won't significantly affect JSON parsing, they will open the door to the next big step in Pydantic validation performance improvements - iterable JSON parser. Currently we parse all JSON, allocating it to vectors, maps and strings; then perform validation on the result. The long term plan is to do validation as we parse JSON which should be a big leap in performance as well as a reduction in memory usage.

---

On msgspec, I don't like commenting on other libraries so I generally avoid saying anything about it, but I feel it's only fair to give my point of view here.

it's an incredibly impressive piece of work from the author, but it goes directly against the security and reliability drive of the whole industry and Python in particular - for example see [this article](https://sethmlarson.dev/security-developer-in-residence-weekly-report-21) - Python is now approved as a memory safe language, in part because you can build extensions in another memory safe language (Rust), and the intention of the PSF is to encourage popular extensions to adopt Rust over C for memory safety, Cryptography is already doing that.

Msgspec in contract is [22,236](https://github.com/jcrist/msgspec/blob/38c3330547087c8f414f4f0b511206e9dd06f97b/msgspec/_core.c) lines of C in one file, maintained effectively by one (albeit very bright and dedicated) person. It's used and audited far less than other projects with similar amounts of C.

If your application really is "critical", msgspec seems to me like a risky choice to me.

**Reply by samuelcolvin** [reply]
https://github.com/pydantic/jiter/pull/55 is merged and I've done a new release of jiter, we're preparing for pydantic-core v2.16.0 release - https://github.com/pydantic/pydantic-core/pull/1165.

This should all be released in pydantic v2.6.0b1 this week.

**Reply by HansBrende** [reply]
@samuelcolvin greatly appreciative of your quick response and fix! Here are my updated benchmarks for pydantic v2.6.0b1:

`MyModel[MyProps].model_validate(json.loads(bytes))` 11.1 s

`MyModel[MyProps].model_validate(orjson.loads(bytes))` 9.1 s

`MyModel[MyProps].model_validate_json(bytes)`: 15.4 s

`my_msgspec_decoder.decode(bytes)`: 1.8 s (this number has decreased from the original 4 seconds I assume due to my enabling the `no_gc` flag on each msgspec struct)

Again, all of these timings are taken as the sum of the individual deserialization times from 210 concurrent server responses.

It seems that there has been quite a large performance enhancement here during validation... however, I'm still getting the _perplexing_ result of `model_validate(json.loads(bytes))` being 37% faster than `model_validate_json(bytes)` (although the gap has indeed narrowed slightly).

I've also tried disabling the GC to see what effect removing gc pauses will have (this is also what timeit does automatically) on the deserialization process. With each execution wrapped between a `gc.disable()` and a `gc.enable()`, we have the following results:

`MyModel[MyProps].model_validate(json.loads(bytes))` 8.1 s     (model_validate: 3.1 s + json.loads: 5.0 s)

`MyModel[MyProps].model_validate(orjson.loads(bytes))` 6.2 s     (model_validate: 3.1 s + orjson.loads: 3.1 s)

`MyModel[MyProps].model_validate_json(bytes)`: 12.3 s

`my_msgspec_decoder.decode(bytes)`: 1.5 s

Removing gc pauses from the equation, the absolute performance is much better for all cases (basically just shaving off 3 seconds for each operation except the msgspec one), but the relative performance of `model_validate_json` becomes even worse, with `model_validate(json.loads)` becoming 52% faster than `model_validate_json` (and `model_validate(orjson.loads)` becoming _twice_ as fast!)

One other detail, which I'm not sure if matters or not... my original pydantic v2.5.3 came via conda, and v2.6.0b1 came via pip.

<sub>**Small clarification on the verbiage in one of my prior comments:** by "critical section", I meant a "critical CPU bottleneck", not "critical" as in "this code is guiding astronauts to the moon."</sub>

**Reply by HansBrende** [reply]
@samuelcolvin @adriangb update: I think I may have spotted where the discrepancy lies. I have a pre-validator on my `Result[P]` model. I didn't mention it earlier because it is doing very minimal work and I (incorrectly) assumed it would add the _same_ amount of work to both `model_validate` and `model_validate_json`. For the purposes of this discussion, you can assume it's implemented as follows:

```python
class Result(BaseModel, Generic[P]):
    id: int
    uuid: UUID
    properties: P
    
    @model_validator(mode="before")
    @classmethod
    def _pre_validate(cls, values):  # no-op prevalidator
        return values
```

Yet somehow, whenever I comment this method out, the performance of `model_validate_json` jumps to be _nearly identical_ to the performance of `model_validate(json.loads)` (although still significantly slower than `model_validate(orjson.loads)`). This is surprising to me, since *both* operations must call this same method if it is defined. I hope this helps!!!

Here are the timings showing the discrepancy between the no-op prevalidator vs. the prevalidator commented out altogether (all run with the GC disabled):

**prevalidator commented out:**
`model_validate + json.loads`: 7.3 s
`model_validate + orjson.loads`: 5.0 s
`model_validate_json`: 7.6 s

**NO-OP prevalidator:**
`model_validate + json.loads`: 7.3 s   (same as before)
`model_validate + orjson.loads`: 5.0 s  (same as before)
`model_validate_json`: 11.2 s   (47% worse!!!)

**Reply by samuelcolvin** [reply]
First of all, it might be worth trying 2.6.0b1 which we released yesterday.

The change in behaviour with a validator is expected - when you add the validator, we have to call [`to_object()`](https://github.com/pydantic/jiter/blob/13070b94e9c24314fe7347fe6f6a6f6d5f468334/src/value.rs#L28-L48) on the `JsonValue` to construct a python object to pass to the function, that involves iterating over all the data being passed to the function. All subsequent validation is then occurring on the python object created by `to_object()`.

If you have a python validator near the root of your model, you'll loose most of the advantage of `model_validate_json`, it might even be slower since constructing a `JsonValue`, then converting that to a python object will always be slower than building a python object directly from JSON as `json.loads` or `orjson` do.

You might also try `pydantic_core.from_json` (as in `model.model_validate(pydantic_core.from_json(data))`) - `from_json` isn't yet faster than `orjson` (although I'm hoping to get there in future), but it's generally faster than `json.loads`, it also takes a keyword argument `cache_strings=False` which might improve performance if your data is big enough to bust the string cache that both `orjson` and `pydantic_core.from_json` use internally.

**Reply by HansBrende** [reply]
@samuelcolvin yep, I saw! (Mentioned at the top of my first comment.) All these numbers are from 2.6.0b1 which I pip-installed this morning.

Following your suggestion, I've added `pydantic_core.from_json` to my timings:

**prevalidator commented out:**
`model_validate + json.loads`: 7.3 s
`model_validate + orjson.loads`: 5.0 s
`model_validate_json`: 7.6 s
`model_validate(pydantic_core.from_json(bytes, cache_strings=True))`: 6.3 s
`model_validate(pydantic_core.from_json(bytes, cache_strings=False))`: 8.2 s

**NO-OP prevalidator:**
`model_validate + json.loads`: 7.3 s (same as before)
`model_validate + orjson.loads`: 5.0 s (same as before)
`model_validate_json`: 11.2 s (47% worse!!!)
`model_validate(pydantic_core.from_json(bytes, cache_strings=True))`: 6.4 s
`model_validate(pydantic_core.from_json(bytes, cache_strings=False))`: 8.3 s

So if validators don't work well with `model_validate_json`, the docs seem a bit misleading here:
![Screenshot 2024-01-20 at 3 48 24 PM](https://github.com/pydantic/pydantic/assets/5846890/5fa12ed7-1fbc-4238-b7b7-184a06000164)
(Or have at least omitted the fact that `model_validate(json.loads(...))` is in fact _way_ better if you *do* have any pre-validators).

Seeing as `model_validate(pydantic_core.from_json(bytes))` actually seems to be faster than `model_validate_json` (in **both** the cases of either having a pre-validator or not)... I guess this brings me back to my original question of what's the point in combining both json-parsing and validation into one step if separating them into two steps is faster?

**Reply by samuelcolvin** [reply]
(sorry I didn't see your first comment before, I've now read it)

Good, point we'll update the documentation. I wasn't aware of cases where `model_validate(json.loads(...))` was faster than `model_validate_json` until this conversation. @sydney-runkle maybe we can do a quick PR to update the performance tips page?

> I guess this brings me back to my original question of what's the point in combining both json-parsing and validation into one step if separating them into two steps is faster?

1. In many cases it is faster faster than two steps
2. There are numerous types that don't really work with JSON in strict mode data unless you're using `model_validate_json` - e.g. `bytes`, `datetime`, `date`, `timedelta`, `tuple`, `set`, `frozenset`, `Path` - any thing other than the 7 primative types defined by JSON - strict mode was a major new feature of Pydantic V2 that I implemented because lots of people had been asking for it, but it only made sense with builtin JSON parsing
3. This is a journey, with many improvements still to come, in particular https://github.com/pydantic/pydantic-core/pull/1085 and iterable JSON parsing that we should be able to start working on once https://github.com/pydantic/pydantic-core/pull/1085 is merged, with those two changes, I'm pretty sure we can reach a point where `model_validate_json` is pretty much always faster than `model_validate(json.loads(...))`.

---

By the way, you might want to be a bit careful about using `orjson` if you're data might have large ints as they're not fully supported - `orjson.loads('1' * 20) != int('1' * 20)`, int parsing should work correctly with `json.laods` and `pydantic_core.from_json`.

**Reply by HansBrende** [reply]
@samuelcolvin Makes sense, thanks for the explanation, and I really appreciate your efforts in this journey! No large integers expected in these payloads, so that shouldn't be a concern. In any case, I've removed `orjson` anyways in favor of `msgspec` from this section of my code (for the time being), as my tests indicate that it does _all_ of this (parsing **and** validation) in a mere **1.5 s** (and I figured out a way to hack it to where I don't _need_ a prevalidator whereas it is unclear how to the same in pydantic).[^1]

As a fellow Rust-enthusiast, I totally understand the complaint about 22K lines of C, however, from a _super practical level_... I **need** that speed in this spot. I'm still using pydantic everywhere else... just not in this one critical CPU bottleneck. That being said, I will be keeping an eye on your performance journey! Theoretically, down the line, it should be possible to get that kind of speed in Rust too I would think.

[^1]: The msgspec hack which removes the need for a prevalidator is declaring `__supertype__: ClassVar[type] = Annotated[str, Meta(extra={'REPLACE_ME': True})]` on a certain `Struct` (the same one I referred to as `IdModel`) so that msgspec thinks `IdModel` is actually a `NewType` around a `str` for parsing purposes, but PyCharm thinks it's a regular `IdModel` for typing purposes. Then, in the `__post_init__` stage, I simply replace these `str` values with a corresponding `IdModel` which I look up _elsewhere_ (don't ask me why 🥴) in the response JSON. I haven't figured out how to do this in pydantic to avoid the prevalidator performance hit, but am open to suggestions and can elaborate if you want more details.

**Comment by PaulRudin** [comment]
JOOI - what's the latest benchmarking here? Is the native pydantic (de)serialization significantly slower than orjson?

**Comment by biozz** [comment]
There is another point, which is kind of discussed here, but not directly - custom encoders.

I have a project, which for the past decade used vanilla `json`, then `ujson`, then `orjson`. This was all before adopting pydantic. There is still quite a lot of legacy code, which uses `dumps` and `loads` methods directly. These methods are abstracted into a separate module (`from app.utils import json`) to be able to change the library for entire project (I know, this is a bold move). I define custom encoders, which were also used in Django models inside `JSONField`. And for each transition from one library to another, I only had to update how encoders are defined, not the way the libraries are used. It was nice to have `json_dumps` and `json_loads` in v1, because I was confident that JSON output would be the same for Pydantic models as it was before introducing Pydantic.

Here are several examples of the encoder rules:

- Decimals are dumped strings
- if Decimal field is None, it becomes `Decimal('0.0')`
- `uuid.UUID` are also dumped as strings
- datetimes are dumped as `isoformat` strings

There are a couple of more  rules, but you get the idea.

I understand, that there is now a concept of [Serialization](https://docs.pydantic.dev/latest/concepts/serialization/#custom-serializers), but that isn't quite the same as before, when I could re-use the well established serialization rules. I guess for now I will have to use `model.dict()` and then use my helper JSON module to serialize the data.

I don't know if I have any specific questions right now, but I just wanted to point out that there is more to the discussion, then performance considerations.

**Reply by samuelcolvin** [reply]
Thanks for this @biozz with v2, you'll probably want to use `model_dump(mode='json')` then apply your custom serialization logic.

I don't know if it's faster or slower than `model_dump_json()`, but in the vast majority of scenarios the performance difference will be negletable.

**Reply by jwhitaker-gridcog** [reply]
We're hitting a dreadful extension of this - we have numpy ints/floats all over our models.
With pydantic 1 we could pass a custom json encoder and call it a day.
With pydantic 2 we can't do this. If we call `model_dump(mode='json')` then we get
```
pydantic_core._pydantic_core.PydanticSerializationError: Unable to serialize unknown type: <class 'numpy.int64'>
```

It's not clear what the escape hatch for us is here.

**Comment by sydney-runkle** [comment]
@phillipuniverse,

Could you please provide the full code snippet used to analyze the differing performance? I'm doing a presentation on performance with pydantic, and want to collect the latest metrics for tests like yours :).

**Reply by phillipuniverse** [reply]
I don’t have a real code snippet analyzing different performance. My question was more about what the migration path is for Pydantic 1.x users using orjson and what the model performance is between 1.x with orjson vs 2.x native.

TBH I’m still not sure of the answer!

**Comment by harshit98** [comment]
@phillipuniverse 

How did you solved it ? Are you still on Pydantic V1 with orjson ? I've similar case but, it's just that I use [msgspec](https://github.com/jcrist/msgspec).

**Comment by triumph1-ably** [comment]
any latest update?

**Comment by samuelcolvin** [comment]
As discussed above, this is not required - pydantic 2 has an extremely fast JSON parser ([jiter](https://github.com/pydantic/jiter)) built in, thus `model_validate_json` is generally (I don't know of an exception to this rule) faster than `model_validate(json.loads(input_json))` or `model_validate(orjson.loads(input_json))`.

Here's an example:

```
model_validate_json(...): 2.94ms
model_validate(json.loads(...)): 4.71ms
model_validate(orjson.loads(...)): 3.24ms
json.loads(...): 2.77ms
pydantic_core.from_json(...): 1.01ms
orjson.loads(...): 1.27ms
```

<details>
<summary>Code</summary>

```py
# /// script
# requires-python = ">=3.13"
# dependencies = [
#     "orjson==3.11.1",
#     "pydantic==2.11.7",
# ]
# ///
import json
import random
import timeit
import uuid
from datetime import date
from typing import Generic, TypeVar

import orjson
import pydantic_core
from pydantic import BaseModel


class PModel(BaseModel):
    field1: str
    field2: str
    field3: int | None = None
    field4: int | None = None
    field5: int | None = None
    field6: int | None = None
    field7: int | None = None
    field8: uuid.UUID | None = None
    field9: uuid.UUID | None = None
    field10: bool | None = None
    field11: int | None = None
    field12: str | None = None
    field13: str | None = None
    field14: bool | None = None
    # field15: IdModel
    id: str
    uuid: uuid.UUID
    field16: str | None = None
    field17: float | None = None
    field18: date | None = None


P = TypeVar('P')


class Result(BaseModel, Generic[P]):
    id: int
    uuid: uuid.UUID
    properties: P


class Response(BaseModel, Generic[P]):
    results: list[Result[P]]


def rand_str():
    return ''.join(random.choice('abcdefghijklmnopqrstuvwxyz') for _ in range(random.randint(1, 1000)))


def rand_num():
    return random.randint(1, 1000)


def gen_p():
    return {
        'field1': rand_str(),
        'field2': rand_str(),
        'field3': rand_num(),
        # skip half of the optional fields
        'field5': rand_num(),
        'field7': rand_num(),
        'field9': str(uuid.uuid4()),
        'field10': True,
        'field12': rand_str(),
        'field14': False,
        'id': rand_str(),
        'uuid': str(uuid.uuid4()),
    }


def gen_result():
    return {
        'id': rand_num(),
        'uuid': str(uuid.uuid4()),
        'properties': gen_p(),
    }


def gen_response():
    return {'results': [gen_result() for _ in range(1000)]}


json_data = json.dumps(gen_response()).encode()
repeat_count = 1_000

timer = timeit.Timer(
    'model_validate_json(json_data)',
    setup='model_validate_json = Response[PModel].model_validate_json',
    globals={'Response': Response, 'PModel': PModel, 'json_data': json_data},
)
timings = timer.repeat(repeat=2, number=repeat_count)
print(f'model_validate_json(...): {timings[-1] / repeat_count * 1_000:.2f}ms')

timer = timeit.Timer(
    'model_validate(json_loads(json_data))',
    setup='model_validate = Response[PModel].model_validate',
    globals={'json_data': json_data, 'json_loads': json.loads, 'Response': Response, 'PModel': PModel},
)
timings = timer.repeat(repeat=2, number=repeat_count)
print(f'model_validate(json.loads(...)): {timings[-1] / repeat_count * 1_000:.2f}ms')

timer = timeit.Timer(
    'model_validate(orjson_loads(json_data))',
    setup='model_validate = Response[PModel].model_validate',
    globals={'json_data': json_data, 'orjson_loads': orjson.loads, 'Response': Response, 'PModel': PModel},
)
timings = timer.repeat(repeat=2, number=repeat_count)
print(f'model_validate(orjson.loads(...)): {timings[-1] / repeat_count * 1_000:.2f}ms')

timer = timeit.Timer('json_loads(json_data)', globals={'json_data': json_data, 'json_loads': json.loads})
timings = timer.repeat(repeat=2, number=repeat_count)
print(f'json.loads(...): {timings[-1] / repeat_count * 1_000:.2f}ms')

timer = timeit.Timer(
    'pydantic_core_from_json(json_data)',
    globals={'json_data': json_data, 'pydantic_core_from_json': pydantic_core.from_json},
)
timings = timer.repeat(repeat=2, number=repeat_count)
print(f'pydantic_core.from_json(...): {timings[-1] / repeat_count * 1_000:.2f}ms')

timer = timeit.Timer('orjson_loads(json_data)', globals={'json_data': json_data, 'orjson_loads': orjson.loads})
timings = timer.repeat(repeat=2, number=repeat_count)
print(f'orjson.loads(...): {timings[-1] / repeat_count * 1_000:.2f}ms')
```

</details>

In addition, orjson has some shortcomings which pydantic_core's JSON parser (jiter) doesn't suffer from - for example orjson parses large ints as floats which looses detail and can break validation:

```py
# /// script
# requires-python = ">=3.13"
# dependencies = [
#     "orjson==3.11.1",
#     "pydantic==2.11.7",
# ]
# ///
import json

import orjson
import pydantic_core
from pydantic import BaseModel


class IntModel(BaseModel):
    value: int


input_json = json.dumps({'value': 2**65})
print('input_json:', input_json)
# > input_json: {"value": 36893488147419103232}
print('IntModel.model_validate_json:', IntModel.model_validate_json(input_json))
# > IntModel.model_validate_json: value=36893488147419103232
print('pydantic_core.from_json:', pydantic_core.from_json(input_json))
# > pydantic_core.from_json: {'value': 36893488147419103232}
print('json.loads:', json.loads(input_json))
# > json.loads: {'value': 36893488147419103232}
print('orjson.loads:', orjson.loads(input_json))
# parses 2**65 as a float!
# > orjson.loads: {'value': 3.6893488147419103e+19}

print('IntModel.model_validate(json.loads(input_json)):', IntModel.model_validate(json.loads(input_json)))
# > IntModel.model_validate(json.loads(input_json)): value=36893488147419103232

print('IntModel.model_validate(orjson.loads(input_json)):', IntModel.model_validate(orjson.loads(input_json)))
"""
pydantic_core._pydantic_core.ValidationError: 1 validation error for IntModel
value
  Unable to parse input string as an integer, exceeded maximum size [type=int_parsing_size, input_value=3.6893488147419103e+19, input_type=float]
    For further information visit https://errors.pydantic.dev/2.11/v/int_parsing_size
"""
```
